src/modules/networks/networks-store.ts

```diff
@@ -1,101 +1,170 @@
 import { Store } from 'store-unit';
-import throttle from 'lodash/throttle';
-import { produce } from 'immer';
-import type { EthereumChainSources } from './Networks';
+import type { EthereumChainConfig } from '../ethereum/chains/ChainConfigStore';
+import type { ChainId } from '../ethereum/transactions/ChainId';
 import { Networks } from './Networks';
-import { get as getNetworks } from './networks-api';
-import { chainsInfo } from './networks-fallback';
+import { getNetworkByChainId, getNetworks } from './networks-api';
+import type { NetworkConfig } from './NetworkConfig';
+import { toNetworkConfig } from './helpers';
 
 interface State {
   networks: Networks | null;
 }
 
+function mergeNetworkConfigs(
+  prevConfigs: NetworkConfig[],
+  nextConfigs: NetworkConfig[]
+) {
+  const nextConfigMap = Object.fromEntries(
+    nextConfigs.map((config) => [config.id, config])
+  );
+  const prevConfigsSet = new Set(prevConfigs.map((item) => item.id));
+  return [
+    ...prevConfigs.map((config) => nextConfigMap[config.id] ?? config),
+    ...nextConfigs.filter((config) => !prevConfigsSet.has(config.id)),
+  ];
+}
+
 export class NetworksStore extends Store<State> {
-  private loaderPromise: Promise<Networks> | null = null;
-  private getEthereumChainSources:
+  private networkConfigs: NetworkConfig[] = [];
+  private customNetworkConfigs: NetworkConfig[] = [];
+  private loaderPromises: Record<string, Promise<Networks>> = {};
+  private getEthereumChainConfigs:
     | null
-    | (() => Promise<EthereumChainSources | undefined>);
-
-  private needsRetry = true;
+    | (() => Promise<EthereumChainConfig[] | undefined>);
 
   constructor(
     state: State,
     {
-      getEthereumChainSources,
+      getEthereumChainConfigs,
     }: {
-      getEthereumChainSources?: NetworksStore['getEthereumChainSources'];
+      getEthereumChainConfigs?: NetworksStore['getEthereumChainConfigs'];
     } = {}
   ) {
     super(state);
-    this.getEthereumChainSources = getEthereumChainSources ?? null;
+    this.getEthereumChainConfigs = getEthereumChainConfigs ?? null;
   }
 
-  private prepareNetworkConfigTypes(
-    ethereumChainSources?: EthereumChainSources
-  ): EthereumChainSources | undefined {
-    if (!ethereumChainSources) {
-      return undefined;
-    }
-    return produce(ethereumChainSources, (draft) => {
-      draft.custom?.ethereumChains?.forEach((chainConfig) => {
-        if (chainConfig.value.native_asset?.decimals) {
-          // there was a problem where some native_token decimals were saved as string
-          chainConfig.value.native_asset.decimals = Number(
-            chainConfig.value.native_asset.decimals
-          );
-        }
-      });
-    });
+  private getKnownIdsSet() {
+    return new Set([
+      ...this.networkConfigs.map((config) => config.id),
+      ...this.customNetworkConfigs.map((config) => config.id),
+    ]);
```

Comment:
Isn't `this.state.networks.collection` already a dictionary with all known ids?

Reply:
It doesn't have it on initial fetchNetworks. I need to think, is it possible to get rid of this helper.

Reply:
Fair

Reply:
See NetworksStore `fetchNetworkById`